File: bacnet-sc-reference-stack-code/dev/src/java/dev/bscs/bacnet/applications/TestSCSC.java
// Copyright, distribution, and usage are defined by the top level LICENSE file.
package dev.bscs.bacnet.applications;

import dev.bscs.bacnet.bacnetsc.SCDatalink;
import dev.bscs.bacnet.bacnetsc.SCProperties;
import dev.bscs.bacnet.bacnetsc.SCVMAC;
import dev.bscs.bacnet.stack.Device;
import dev.bscs.common.*;

/**
 * An application that creates one device that is a router between BACnet/SC and BACnet/SC and can optionally be hubs.
 * The configuration of this application is controlled by the TestSCSC.properties file. See the top level README for details.
 * Also see {@link Application} class for details of the structure of the meaning of these overridden methods.
 * @author drobin
 */
public class TestSCSC extends Application {

    private static SCDatalink scDatalink1;
    private static SCDatalink scDatalink2;

    @Override protected void start() {
        Device device = new Device(configuration);
        int scNetworkNumber1 = configuration.getInteger("app.scNetworkNumber1", 55501);
        int scNetworkNumber2 = configuration.getInteger("app.scNetworkNumber2", 55502);

        SCProperties scProperties1 = new SCProperties(configuration);

        SCProperties scProperties2 = new SCProperties(configuration);

        scProperties2.vmac = new SCVMAC(configuration.getMAC("app.2.sc.vmac", 6, new byte[6]));
        scProperties2.primaryHubURI = configuration.getString("app.2.sc.primaryHubURI", "");
        scProperties2.hubFunctionEnable = configuration.getBoolean("app.2.sc.hubFunctionEnable", true);
        scProperties2.hubFunctionBindURI = configuration.getString("app.2.sc.hubFunctionBindURI", "");

        scDatalink1 = new SCDatalink("SC-1", scProperties1, device, scNetworkNumber1);
        scDatalink2 = new SCDatalink("SC-2", scProperties2, device, scNetworkNumber2);

        scDatalink1.start();
        scDatalink2.start();
    }

    @Override protected void addCommands() {
        BACnetCommands.addCommands();
        SCCommands.addCommands();
    }

    @Override protected void stop() {
        scDatalink1.stop();
        scDatalink2.stop();
    }

    @Override protected void close() {
        scDatalink1.close();
        scDatalink2.close();
    }

}

File: bacnet-sc-reference-stack-code/scripts/TestSCSC.py
from dev.bscs.common import Shell
from dev.bscs.common import Timer
from dev.bscs.common import Command
from dev.bscs.common import Application
from dev.bscs.common import CommandProcessor
from dev.bscs.common import Formatting
from dev.bscs.bacnet.applications import BACnetCommands
from dev.bscs.bacnet.applications import SCCommands
from dev.bscs.bacnet.stack import AuthData
from dev.bscs.bacnet.bacnetsc import SCProperties
from dev.bscs.bacnet.bacnetsc import SCDatalink
from dev.bscs.bacnet.bacnetsc import SCVMAC
from dev.bscs.bacnet.stack import Binding
from dev.bscs.bacnet.stack import Device
from dev.bscs.bacnet.stack import Failure
from dev.bscs.bacnet.stack.data import BACnetPropertyIdentifier
from dev.bscs.bacnet.stack.data.base import BACnetData
from dev.bscs.bacnet.stack.data.base import BACnetObjectIdentifier
from dev.bscs.bacnet.stack.services import ReadPropertyClient
from dev.bscs.bacnet.stack.services import WhoIsClient
from dev.bscs.events import EventListener
from dev.bscs.events import EventLoop
from dev.bscs.events import EventType
from java.io import FileNotFoundException
from java.io import FileReader
from java.io import IOException
from java.io import Reader
from java.lang import String
from java.util import Queue
from java.util.concurrent import ConcurrentLinkedQueue
from jarray import array
	
# In addition to running scripts from the terminal or console, the entire 
# application can be written as a python class that extends the java Application 
# class.  This allows the python code to install event handlers and callbacks to 
# execute arbitrary scripted actions without user interaction after the start() 
# method returns. A python application can be created by specifying 
# "common.mainScript" in addition to the "common.mainClass" (which in that case 
# indicates the name of the python class). e.g., 
#    common.mainClass   = TestSCSC
#    common.mainScript  = scripts/TestSCSC.py
 
class TestSCSC(Application):

    def start(self):
        configuration = Application.configuration
        device = Device(configuration)
        scNetworkNumber1 = configuration.getInteger("app.scNetworkNumber1", 55501)
        scNetworkNumber2 = configuration.getInteger("app.scNetworkNumber2", 55502)

        scProperties1 = SCProperties(configuration)

        scProperties2 = SCProperties(configuration)

        scProperties2.vmac = SCVMAC(configuration.getMAC("app.2.sc.vmac", 6, array([0,0,0,0,0,0],'b')))
        scProperties2.primaryHubURI = configuration.getString("app.2.sc.primaryHubURI", "")
        scProperties2.hubFunctionEnable = configuration.getBoolean("app.2.sc.hubFunctionEnable", True)
        scProperties2.hubFunctionBindURI = configuration.getString("app.2.sc.hubFunctionBindURI", "")

        self.scDatalink1 = SCDatalink("SC-1", scProperties1, device, scNetworkNumber1)
        self.scDatalink2 = SCDatalink("SC-2", scProperties2, device, scNetworkNumber2)

        self.scDatalink1.start()
        self.scDatalink2.start()
    
    def addCommands(self):
        BACnetCommands.addCommands()
        SCCommands.addCommands()
    
    def stop(self):
        self.scDatalink1.stop()
        self.scDatalink2.stop()
    
    def close(self):
        self.scDatalink1.close()
        self.scDatalink2.close()
        

